### ISSUE TITLE
Inconsistencies in ServiceWorkerClients get and matchAll implementations

### ISSUE THREAD

**Issue by youennf** [w3c/ServiceWorker#1734]
I had a look at Chrome, Firefox and Safari implementation of ServiceWorkerClients get and matchAll implementations.
They are not really aligned for fetch event resultingClientId.

For ServiceWorkerClients.get called synchronously using resultingClientId from a navigation fetch event handler:
1. Chrome is waiting for the document to be parsed before resolving the get promise. This is shown in https://wpt.fyi/results/service-workers/service-worker/controlled-iframe-postMessage.https.html where the fetch response is waiting on `get` promise to resolve.
2. Firefox and Safari are resolving the promise right away (I believe Firefox exposes the client URL `about:blank` and Safari exposes the request URL)

For ServiceWorkerClients.matchAll called synchronously from a navigation fetch event handler
1. Chrome is not exposing the resultingClientId client
2. Firefox is exposing the resultingClientId client (url is `about:blank`)
3. Shipping Safari is exposing the resultingClientId client (url is the request URL), though WebKit ToT recently changed to match Chrome to fix some potential issues in case of push event processing.

@asutherland, @yoshisatoyanagisawa, thoughts?

**Comment by youennf** [comment]
Digging on `matchAll`, it seems that setting ` includeUncontrolled: true` will align Chrome's results with shipping safari.

Digging on `get`, it seems from testing that Chrome will wait for the service worker to provide a response object, but will not wait to start parsing the response body.

Only Firefox is exposing `about:blank` URLs.
If calling `matchAll` or `get`, the client URL will, at some point migrate from `about:blank` to the fetch event request URL.

**Comment by yoshisatoyanagisawa** [comment]
> Chrome is waiting for the document to be parsed before resolving the get promise.

Yes.  Chrome wait until ServiceWorkerClient is execution ready.
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_version.cc;l=1700;drc=fc7ddc185b7f027dd7f4981ba8b2334c69a2e79c
It should be set soon after the commit navigation IPC is sent from the browser process.  I understand the timing is just after getting all HTTP headers.
https://source.chromium.org/chromium/chromium/src/+/main:content/renderer/service_worker/service_worker_network_provider_for_frame.cc;l=48;drc=a45502c46d75f210c783e07384138379ea1e46e4

> Chrome is not exposing the resultingClientId client

Chrome is explicitly excluding it.
No other controllee case:
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_client_utils.cc;l=609;drc=a45502c46d75f210c783e07384138379ea1e46e4
With other controllee case:
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_client_utils.cc;l=617;drc=a45502c46d75f210c783e07384138379ea1e46e4

What I am a bit confused for Chrome implementation is that `SetControllerRegistration` is called before dispatching the fetch handler.
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_controllee_request_handler.cc;l=456;drc=a45502c46d75f210c783e07384138379ea1e46e4
Then, if there is other clients, will we see the client with the resulting client id?

@hiroshige-g recently touched the code around this area.  Let me listen to your thoughts.

**Comment by asutherland** [comment]
@youennf The Firefox code is definitely trying to wait for the clients to be execution ready in both cases:
- For matchAll [we exclude clients that are not execution ready](https://searchfox.org/mozilla-central/rev/5f2c1701846a54c484d7dd46a291b796f5a67cac/dom/clients/manager/ClientManagerService.cpp#538-540)
- For get [Firefox waits for the execution ready promise](https://searchfox.org/mozilla-central/rev/5f2c1701846a54c484d7dd46a291b796f5a67cac/dom/clients/manager/ClientManagerService.cpp#686-690) to resolve.

How are you testing this?

**Comment by asutherland** [comment]
> Would it help if I upload draft WPT tests to make this clearer?

Yes, that would be great, thank you  Then I can run them under pernosco and figure out what's going on.

**Comment by youennf** [comment]
I uploaded https://github.com/web-platform-tests/wpt/pull/49393 to make progress on this issue.

**Comment by youennf** [comment]
@asutherland, did you get a chance to look at the test?
